pkg/engine/lifecycle_test.go

```diff
@@ -5216,6 +5216,197 @@ func TestSingleResourceDefaultProviderGolangLifecycle(t *testing.T) {
 	p.Run(t, nil)
 }
 
+// Inspired by transformations_test.go.
+func TestSingleResourceDefaultProviderGolangTransformations(t *testing.T) {
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				CreateF: func(urn resource.URN,
+					news resource.PropertyMap, timeout float64) (resource.ID, resource.PropertyMap, resource.Status, error) {
+
+					return "created-id", news, resource.StatusOK, nil
+				},
+				ReadF: func(urn resource.URN, id resource.ID,
+					inputs, state resource.PropertyMap) (plugin.ReadResult, resource.Status, error) {
+					return plugin.ReadResult{Inputs: inputs, Outputs: state}, resource.StatusOK, nil
+				},
+			}, nil
+		}),
+	}
+
+	newResource := func(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) error {
+		var res testResource
+		return ctx.RegisterResource("pkgA:m:typA", name, &testResourceInputs{
+			Foo: pulumi.String("bar"),
+		}, &res, opts...)
+	}
+
+	newComponent := func(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) error {
+		var res testResource
+		err := ctx.RegisterComponentResource("pkgA:m:typA", name, &res, opts...)
+		if err != nil {
+			return err
+		}
+
+		var resChild testResource
+		return ctx.RegisterResource("pkgA:m:typA", name+"Child", &testResourceInputs{
+			Foo: pulumi.String("bar"),
+		}, &resChild, pulumi.Parent(&res))
+	}
+
+	program := deploytest.NewLanguageRuntime(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		ctx, err := pulumi.NewContext(context.Background(), pulumi.RunInfo{
+			Project:     info.Project,
+			Stack:       info.Stack,
+			Parallel:    info.Parallel,
+			DryRun:      info.DryRun,
+			MonitorAddr: info.MonitorAddress,
+		})
+		assert.NoError(t, err)
+
+		return pulumi.RunWithContext(ctx, func(ctx *pulumi.Context) error {
+			// Scenario #1 - apply a transformation to a CustomResource
+			res1Transformation := func(args *pulumi.ResourceTransformationArgs) *pulumi.ResourceTransformationResult {
+				return &pulumi.ResourceTransformationResult{
+					Props: args.Props,
+					Opts:  append(args.Opts, pulumi.AdditionalSecretOutputs([]string{"output"})),
```

Comment:
Ideally we would use a `mergeOptions` style API here - but I think we don’t have it yet in Go. Not critical - but possibly worth a TODO pointing at the issue tracking that.